pkg/ctl/install/pki.go

```diff
@@ -0,0 +1,167 @@
+	clientCertificate string
+	clientKey         string
+}
+
+type publicKeyInfrastructure struct {
+	caCertificate     []byte
+	caKey             []byte
+	serverCertificate []byte
+	serverKey         []byte
+	clientCertificate []byte
+	clientKey         []byte
+}
+
+type devnull struct{}
+
+func (devnull) Debug(string)   {}
+func (devnull) Info(string)    {}
+func (devnull) Warning(string) {}
+func (devnull) Err(string)     {}
+func (devnull) Crit(string)    {}
+func (devnull) Emerg(string)   {}
+func init() {
+	// Disable logging of cfssl since we handle errors explicitly
+	log.SetLogger(devnull{})
+}
+
+func newPKI(hostName string, validFor time.Duration, rsaKeyBitSize int) (*publicKeyInfrastructure, error) {
+	keyReq := &csr.KeyRequest{
+		A: "rsa",
+		S: rsaKeyBitSize,
+	}
+
+	// Generate CA
+	caReq := &csr.CertificateRequest{
+		KeyRequest: keyReq,
+		CN:         "CA",
+	}
+	caCert, _, caKey, err := initca.New(caReq)
+	if err != nil {
+		return nil, fmt.Errorf("cannot generate root CA: %s", err)
+	}
+
+	// Generate Server Certificate
+	serverCert, serverKey, err := generateCertificate(caCert, caKey, keyReq, hostName, "server", validFor)
+	if err != nil {
+		return nil, fmt.Errorf("cannot generate server certificate: %s", err)
+	}
+
+	// Generate Client certificate
+	generateCertificate(caCert, caKey, keyReq, hostName, "client", validFor)
+	clientCert, clientKey, err := generateCertificate(caCert, caKey, keyReq, hostName, "client", validFor)
+	if err != nil {
+		return nil, fmt.Errorf("cannot generate client certificate: %s", err)
+	}
+	pki := &publicKeyInfrastructure{
+		caCertificate:     caCert,
+		caKey:             caKey,
+		serverCertificate: serverCert,
+		serverKey:         serverKey,
+		clientCertificate: clientCert,
+		clientKey:         clientKey,
+	}
+	return pki, nil
+
+}
+
+func generateCertificate(caCert []byte, caKey []byte, keyReq *csr.KeyRequest,
+	hostName string, commonName string, validFor time.Duration) ([]byte, []byte, error) {
+	policy := &config.Signing{
+		Default: &config.SigningProfile{
+			Expiry: validFor,
+			Usage:  []string{"signing", "key encipherment", "server auth", "client auth"},
+		},
+	}
+	parsedCa, err := helpers.ParseCertificatePEM(caCert)
+	if err != nil {
+		return nil, nil, fmt.Errorf("malformed generated private CA certificate: %s", err)
+	}
+	priv, err := helpers.ParsePrivateKeyPEM(caKey)
+	if err != nil {
+		return nil, nil, fmt.Errorf("malformed generated private CA key: %s", err)
+	}
+	certSigner, err := local.NewSigner(priv, parsedCa, signer.DefaultSigAlgo(priv), policy)
+	if err != nil {
+		return nil, nil, fmt.Errorf("cannot create signer: %s", err)
+	}
+	serverReq := &csr.CertificateRequest{
+		KeyRequest: keyReq,
+		CN:         commonName,
+		Hosts:      []string{hostName},
+	}
+	g := &csr.Generator{Validator: genkey.Validator}
+	csrBytes, key, err := g.ProcessRequest(serverReq)
+	if err != nil {
+		return nil, nil, fmt.Errorf("cannot process server-certificate signing request: %s", err)
```

Comment:
```suggestion
		return nil, nil, fmt.Errorf("cannot process server certificate signing request: %s", err)
```

Reply:
Also, are you sure these errors need to be wrapped? Maybe it'd be okay to just return whatever the library returns, and wrap the result to say that it's to do with Tiller certificate generation?

Reply:
I would rather add context

Reply:
Do you know what the errors say?

Having errors that say the same things twice or more is a extremely annoying.

Reply:
I don't think it says the same thing twice. Find a counter-example and I will fix it.